subsys/net/lib/dhcpv4/dhcpv4_server.c

```diff
@@ -667,13 +683,175 @@ static uint32_t dhcpv4_get_lease_time(uint8_t *options, uint8_t optlen)
 	return CONFIG_NET_DHCPV4_SERVER_ADDR_LEASE_TIME;
 }
 
+#if defined(DHCPV4_SERVER_ICMP_PROBE)
+static int dhcpv4_probe_address(struct dhcpv4_server_ctx *ctx,
+				 struct dhcpv4_addr_slot *slot)
+{
+	struct sockaddr_in dest_addr = {
+		.sin_family = AF_INET,
+		.sin_addr = slot->addr,
+	};
+	int ret;
+
+	ret = net_icmp_send_echo_request(&ctx->probe_ctx.icmp_ctx, ctx->iface,
+					 (struct sockaddr *)&dest_addr,
+					 NULL, ctx);
+	if (ret < 0) {
+		LOG_ERR("Failed to send ICMP probe");
+	}
+
+	return ret;
+}
+
+static int echo_reply_handler(struct net_icmp_ctx *icmp_ctx,
+			      struct net_pkt *pkt,
+			      struct net_icmp_ip_hdr *ip_hdr,
+			      struct net_icmp_hdr *icmp_hdr,
+			      void *user_data)
+{
+	struct dhcpv4_server_ctx *ctx = user_data;
+	struct dhcpv4_server_probe_ctx *probe_ctx = &ctx->probe_ctx;
+	struct dhcpv4_addr_slot *new_slot = NULL;
+	struct in_addr peer_addr;
+
+	ARG_UNUSED(icmp_ctx);
+	ARG_UNUSED(pkt);
+	ARG_UNUSED(ip_hdr);
+	ARG_UNUSED(icmp_hdr);
+
+	k_mutex_lock(&server_lock, K_FOREVER);
+
+	if (probe_ctx->slot == NULL) {
+		goto out;
+	}
+
+	if (ip_hdr->family != AF_INET) {
+		goto out;
+	}
+
+	net_ipv4_addr_copy_raw((uint8_t *)&peer_addr, ip_hdr->ipv4->src);
+	if (!net_ipv4_addr_cmp(&peer_addr, &probe_ctx->slot->addr)) {
+		goto out;
+	}
+
+	LOG_DBG("Got ICMP probe response, blocking address %s",
+		net_sprint_ipv4_addr(&probe_ctx->slot->addr));
+
+	probe_ctx->slot->state = DHCPV4_SERVER_ADDR_DECLINED;
```

Comment:
We should have some mechanism to clear this flag, no? Else we might run out of the pool if a flaky client blocks all IPs (hypothetically)? 

One approach would be to clear all declined flags due to probing if the pool is full.

Reply:
RFC is not very clear about this, in case of Decline message it just says that the server should mark the address as no longer available, for probing there's no info whatsoever.

I can check how it's resolved in for example Linux implementation of the DHCP server, maybe some timeout or as you say reusing the oldest declined entry could be done. But I suggest we handle this in a separate PR.